File: clients/desktop/Core/PairingHost.cs
using System;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;
using Makaretu.Dns;

namespace Mozz.Desktop.Core;

/// <summary>
/// A fresh desktop waiting to join an existing circle.
/// </summary>
/// <remarks>
/// This is the reverse of <see cref="PairingDiscovery"/>. A desktop with no
/// circle listens and advertises; an established phone or computer opens
/// Devices, discovers it, and connects. That makes setup order irrelevant:
/// phone first and desktop first are the same ceremony with the roles reversed.
/// </remarks>
public sealed class PairingHost : IDisposable
{
    private readonly TcpListener _listener;
    private readonly ServiceDiscovery _discovery;
    private readonly ServiceProfile _profile;

    private PairingHost(string name)
    {
        _listener = new TcpListener(IPAddress.Any, 0);
        _listener.Start();
        var port = ((IPEndPoint)_listener.LocalEndpoint).Port;

        _discovery = new ServiceDiscovery();
        _profile = new ServiceProfile(
            name,
            PairingDiscovery.ServiceType,
            checked((ushort)port));
        _discovery.Advertise(_profile);
        _discovery.Announce(_profile);
    }

    public static PairingHost Start(string name) => new(name);

    internal int Port => ((IPEndPoint)_listener.LocalEndpoint).Port;

    public async Task<PairingLink> AcceptAsync(CancellationToken token = default)
    {
        var client = await _listener.AcceptTcpClientAsync(token).ConfigureAwait(false);
        return PairingLink.FromAcceptedClient(client);
    }

    public void Dispose()
    {
        _discovery.Unadvertise(_profile);
        _discovery.Dispose();
        _listener.Stop();
    }
}

File: clients/android/app/src/main/java/com/thatcube/mozz/pairing/PairingHost.kt
package com.thatcube.mozz.pairing

import android.content.Context
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.Closeable
import java.net.ServerSocket

/**
 * A phone with no circle, waiting to be let into one.
 *
 * The reverse of [PairingDiscovery]. The device asking to be admitted listens
 * and advertises; an established device opens its own Devices screen, finds
 * this one and connects. That makes setup order irrelevant — phone first and
 * desktop first are the same ceremony with the roles swapped.
 */
class PairingHost private constructor(
    private val manager: NsdManager,
    private val socket: ServerSocket,
    private val registration: NsdManager.RegistrationListener,
) : Closeable {

    val port: Int get() = socket.localPort

    /** Blocks until the other device connects. */
    suspend fun accept(): PairingLink = withContext(Dispatchers.IO) {
        PairingLink(socket.accept())
    }

    override fun close() {
        runCatching { manager.unregisterService(registration) }
        runCatching { socket.close() }
    }

    companion object {
        /** The same service the desktop advertises and the iPhone browses for. */
        const val SERVICE_TYPE = "_mozz._tcp"

        fun start(context: Context, name: String): PairingHost {
            val manager = context.getSystemService(Context.NSD_SERVICE) as NsdManager
            // Port 0: the system picks one and tells us, which is then what goes
            // in the advertisement. A fixed port would collide with a second
            // Mozz on the same device and, worse, with anything else.
            val socket = ServerSocket(0)
            val info = NsdServiceInfo().apply {
                serviceName = name
                serviceType = SERVICE_TYPE
                port = socket.localPort
            }
            val registration = object : NsdManager.RegistrationListener {
                override fun onServiceRegistered(info: NsdServiceInfo) = Unit
                override fun onRegistrationFailed(info: NsdServiceInfo, errorCode: Int) {
                    // The only breadcrumb when a phone never shows up in the
                    // other device's list. Nothing else reports this.
                    Log.w(TAG, "could not advertise for pairing (error $errorCode)")
                }
                override fun onServiceUnregistered(info: NsdServiceInfo) = Unit
                override fun onUnregistrationFailed(info: NsdServiceInfo, errorCode: Int) = Unit
            }
            manager.registerService(info, NsdManager.PROTOCOL_DNS_SD, registration)
            return PairingHost(manager, socket, registration)
        }

        private const val TAG = "MozzPairing"
    }
}
